crates/matrix-sdk-ui/src/room_list/mod.rs

```diff
@@ -282,10 +285,21 @@ impl RoomListService {
 
     /// Get a [`Room`] if it exists.
     pub async fn room(&self, room_id: &RoomId) -> Result<Room, Error> {
-        match self.sliding_sync.get_room(room_id).await {
-            Some(room) => Room::new(self.sliding_sync.clone(), room).await,
-            None => Err(Error::RoomNotFound(room_id.to_owned())),
+        {
+            let rooms = self.rooms.read().await;
+            if let Some(room) = rooms.get(room_id) {
+                return Ok(room.clone());
+            }
         }
+
+        let room = match self.sliding_sync.get_room(room_id).await {
+            Some(room) => Room::new(self.sliding_sync.clone(), room)?,
+            None => return Err(Error::RoomNotFound(room_id.to_owned())),
+        };
+
+        self.rooms.write().await.insert(room_id.to_owned(), room.clone());
```

Comment:
If a room has been left, yeah it should be removed. Ivan has ideas that involve using a `RingBuffer` instead here, for a limited amount of `Room`s, but that limit would make this fix irrelevant.

We're spamming the embedder with `Filled` events for every room, after a background -> foreground transition, and that's likely the root cause to fix.

Reply:
The only clean solution I can think of, is to remove `RoomListEntry::Invalidate` entirely, thus reducing the number of `Invalidate` and `Filled` updates that are sent to the user. With that, combined to a `latest_event` stored inside `RoomInfo` (cf #2085), it would help to remove this cache entirely.

Let's consider this as a hot fix. I would add a `TODO` note though.